openedx/core/djangoapps/course_groups/models.py

```diff
@@ -58,6 +59,99 @@ def create(cls, name, course_id, group_type=COHORT):
         )
 
 
+class CohortMembership(models.Model):
+    """Used internally to enforce our particular definition of uniqueness"""
+
+    course_user_group = models.ForeignKey(CourseUserGroup)
+    user = models.ForeignKey(User)
+    course_id = CourseKeyField(max_length=255)
+
+    previous_cohort = None
+    previous_cohort_name = None
+    previous_cohort_id = None
+
+    class Meta(object):
+        unique_together = (('user', 'course_id'), )
+
+    # The sole purpose of overriding this method is to get the django 1.6 behavior of allowing 'validate_unique'
+    # For django 1.8 upgrade, just remove this method and allow the base method to be called instead.
+    # Reference: https://docs.djangoproject.com/en/1.6/ref/models/instances/, under "Validating Objects"
+    def full_clean(self, **kwargs):
+        self.clean_fields()
+        self.clean()
+        if 'validate_unique' not in kwargs or kwargs['validate_unique'] is True:
+            self.validate_unique()
+
+    def clean_fields(self, *args, **kwargs):
+        if self.course_id is None:
+            self.course_id = self.course_user_group.course_id
+        super(CohortMembership, self).clean_fields(*args, **kwargs)
+
+    def clean(self):
+        if self.course_user_group.group_type != CourseUserGroup.COHORT:  # pylint: disable=E1101
+            raise ValidationError("CohortMembership cannot be used with CourseGroup types other than COHORT")
+        if self.course_user_group.course_id != self.course_id:
+            raise ValidationError("Non-matching course_ids provided")
+
+    def save(self, *args, **kwargs):
+        # Avoid infinite recursion if creating from get_or_create() call below.
+        if 'force_insert' in kwargs and kwargs['force_insert'] is True:
+            super(CohortMembership, self).save(*args, **kwargs)
+            return
+
+        self.full_clean(validate_unique=False)
+
+        # This loop has been created to allow for optimistic locking, and retrial in case of losing a race condition.
+        # The limit is 2, since select_for_update ensures atomic updates. Creation is the only possible race condition.
+        max_retries = 2
+        success = False
+        for dummy_counter in range(max_retries):
+            saved_membership = None
```

Comment:
Nit: is there a reason why you set saved_membership to None originally? You don't do this for created, and I'm having trouble understanding the difference.